crates/telio-firewall/src/firewall.rs

```diff
@@ -704,30 +743,115 @@ impl StatefullFirewall {
         };
 
         let identifier_sequence = {
-            let payload = icmp_packet.payload();
-            if payload.len() < 4 {
-                return None;
-            }
-            #[allow(index_access_check)]
-            let bytes = payload[0..4].try_into();
-            if let Ok(bytes) = bytes {
-                u32::from_ne_bytes(bytes)
-            } else {
-                return None;
-            }
+            let bytes = icmp_packet
+                .payload()
+                .get(0..4)
+                .and_then(|b| b.try_into().ok());
+            let bytes = unwrap_option_or_return!(bytes, None);
+            u32::from_ne_bytes(bytes)
         };
         let (remote_addr, local_addr) = if inbound {
             (ip.get_source().into(), ip.get_destination().into())
         } else {
             (ip.get_destination().into(), ip.get_source().into())
         };
-        Some(IcmpConn {
+        let key = IcmpConn {
             remote_addr,
             local_addr,
             response_type,
             identifier_sequence,
+        };
+        Some(IcmpKey::IcmpConn(key))
+    }
+
+    fn build_icmp_error_key(icmp_packet: IcmpPacket, is_v4: bool) -> Option<IcmpKey> {
+        let inner_packet = unwrap_option_or_return!(icmp_packet.payload().get(4..), None);
+        let (udp_key, tcp_key, icmp_key) = if is_v4 {
```

Comment:
I would love to make this generic so that the if and else blocks don't need to be duplicated when the only difference is a single type change, `Ipv4Packet` and `Ipv6Packet`, but due to lifetimes I couldn't make it happen without resorting to a macro, and I don't want to introduce that for this small case